torch/csrc/jit/script/compiler.cpp

```diff
@@ -1499,11 +1553,27 @@ struct to_ir {
 
   std::shared_ptr<SugaredValue> emitApplyExpr(Apply &apply, size_t n_binders) {
     auto sv = emitSugaredExpr(apply.callee(), 1);
-    auto inputs = getNamedValues(apply.inputs(), true);
     auto attributes = fmap(apply.attributes(), [&](const Attribute& attr) {
       return NamedValue(attr.range(), attr.name().name(), emitExpr(attr.value()));
     });
-    return sv->call(apply.callee().range(), method, inputs, attributes, n_binders);
+
+    auto loc = apply.callee().range();
+    if (sv->kind() == "fork") {
+      JIT_ASSERT(attributes.size() == 0);
+      JIT_ASSERT(n_binders == 1);
+      auto& trees = apply.inputs().tree()->trees();
+      if (trees.size() != 2) {
+        throw ErrorReport(loc) << "Expected a module and a single tuple argument to fork()";
+      }
+
+      auto sugared_expr = emitSugaredExpr(Expr(trees[0]), 1);
+      TreeList sliced_trees(trees.begin() + 1, trees.end());
+      auto& input = getNamedValues(sliced_trees, true)[0];
+      return static_cast<ForkValue*>(sv.get())->call(loc, method, sugared_expr, input);
+    } else {
+      auto inputs = getNamedValues(apply.inputs(), true);
+      return sv->call(loc, method, inputs, attributes, n_binders);
```

Comment:
[generating the function] It is important that we generate the call to the forked function in a way that (1) supports all the callable things in TorchScript and (2) is a pure function of its inputs. There are methods in TorchScript so the arguments to a call do not always contain all the true arguments to the functions. I suggest this approach to get this correct:

1. initially create a sub-block, as is already done. Do not add any input arguments to the block, they are unnecessary at this point. The `schema()` function is not required.
2. change the insert point to the block and do this code similar to before (note sliced_inputs is NamedValues for all but the first value in inputs):
```
auto fn_sugared_output = forked->call(loc, m, sliced_inputs, attributes, 1);
auto fn_simple_output = fn_sugared_output->asValue(loc, m);
body_block->registerOutput(fn_simple_output);
```
3. "Lambda lift" the block into its own Graph. That is, for each unique value used in the block that is defined outside of the block, add it as a formal input to the Graph, and then copy the body of the block into the new Graph, replacing the actual inputs with the formal ones. Similarly, add the actual value (which corresponds to the formal input) as an input to the fork primitive. We can go over this in person if it isn't immediately clear.

4. add the Graph as an attribute of the Fork node. The implementation of Fork should then construct an Interpreter object for this graph and use it to calculate the actual value.